bin/light-base/src/runtime_service.rs

```diff
@@ -593,98 +208,102 @@ impl<TPlat: Platform> RuntimeService<TPlat> {
 
+        // Extract the components of the `FinalizedBlockRuntimeKnown`. We are guaranteed by the
+        // block above to be in this state.
+        let (tree, finalized_block, pinned_blocks, all_blocks_subscriptions) =
+            match &mut guarded_lock.tree {
+                GuardedInner::FinalizedBlockRuntimeKnown {
+                    tree,
+                    finalized_block,
+                    pinned_blocks,
+                    all_blocks_subscriptions,
+                } => (
+                    tree,
+                    finalized_block,
+                    pinned_blocks,
+                    all_blocks_subscriptions,
+                ),
+                _ => unreachable!(),
+            };
+
         let (mut tx, new_blocks_channel) = mpsc::channel(buffer_size);
         let subscription_id = guarded_lock.next_subscription_id;
         guarded_lock.next_subscription_id += 1;
 
-        let non_finalized_blocks_ancestry_order: Vec<_> = match &guarded_lock.tree {
-            GuardedInner::FinalizedBlockRuntimeKnown {
-                tree,
-                finalized_block,
-            } => {
-                let decoded_finalized_block =
-                    header::decode(&finalized_block.scale_encoded_header).unwrap();
+        let decoded_finalized_block =
+            header::decode(&finalized_block.scale_encoded_header).unwrap();
+        pinned_blocks.insert(
+            (subscription_id, finalized_block.hash),
+            (
+                tree.finalized_async_user_data().clone(),
+                *decoded_finalized_block.state_root,
+                decoded_finalized_block.number,
+            ),
+        );
+
+        let non_finalized_blocks_ancestry_order: Vec<_> = tree
+            .input_iter_ancestry_order()
+            .filter_map(|block| {
+                let runtime = block.async_op_user_data?.clone();
+                let block_hash = block.user_data.hash;
+                let parent_runtime = tree
+                    .parent(block.id)
+                    .map_or(tree.finalized_async_user_data().clone(), |parent_idx| {
+                        tree.block_async_user_data(parent_idx).unwrap().clone()
+                    });
+
+                let parent_hash = *header::decode(&block.user_data.scale_encoded_header)
+                    .unwrap()
+                    .parent_hash; // TODO: correct? if yes, document
+                debug_assert!(
+                    parent_hash == finalized_block.hash
+                        || tree
+                            .input_iter_ancestry_order()
+                            .any(|b| parent_hash == b.user_data.hash
+                                && b.async_op_user_data.is_some())
+                );
 
-                guarded_lock.pinned_blocks.insert(
-                    (subscription_id, finalized_block.hash),
+                let decoded_header = header::decode(&block.user_data.scale_encoded_header).unwrap();
+                pinned_blocks.insert(
```

Comment:
for some reason mutating the state while collecting items feels bad, but can't find a good reason why. feel free to ignore

Reply:
Ah I agree, that's an accident (wasn't the case before the PR) and I'll change that.